woocommerce/woocommerce-gutenberg-products-block
# Create withQueryStringValues HOC and use it in ProductGrid component

This PR refactors the HOCs used by the `<ProductGrid />` block into a single one named `withQueryStringValues` which keeps the state in sync with the URL query strings.

@nerrad this PR undos some of the changes we did in previous PRs (we are no longer using any of the `withBrowserWindowProp` HOCs, I found them not to play well when we needed to use `window` properties inside lifecycle methods, that brought me to extract all the logic into another HOC, but would love to hear your thoughts on this approach.

If there is no window object, `withQueryStringValues` will still work as expected but it won't update/read the query string values from the URL.

### How to test the changes in this Pull Request:

1. Create a post with more than one _All Products_ block.
2. Preview it and change the sort option and navigate to another page.
3. Verify it works as expected.
4. Try navigating back/forward with your browser and verify the sort options/pages are updated accordingly and without causing a page refresh.
5. Modify [these lines](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1040/files#diff-c7272801d58b2ba537ca2909a3e02619R13) and reload the page:
```diff
-		hasWindowDependencies =
-			typeof window === 'object' &&
-			window.hasOwnProperty( 'history' ) &&
-			window.hasOwnProperty( 'location' ) &&
-			typeof window.addEventListener === 'function' &&
-			typeof window.removeEventListener === 'function';
+		hasWindowDependencies = false;
```
6. Verify you can still change the sort option and navigate to another page even though the URL is no longer updated.
